### PR TITLE
fix(dflash): remove incorrect padding shift and fix cross-block data leak

### LINKED ISSUE #470: [Bug] dflash cross block data leak
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/SpecForge/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

I tried training a dflash model using the script in this repo, and noticed that performance is much better when `num-anchors` is small. When `num-anchors` is large, the training loss/acc look abnormally good (loss unusually low / acc unusually high compared to expect), but all the evaluation results are much worse than reported in its paper.

So I dug into the code and found the `random-anchor` path: cross-block context visibility is determined purely by the `block index` (`k_ctx < q_block_id`). When the anchor spacing becomes `smaller than `block_size`, blocks can overlap. This overlap allows a later block to see future target hidden states from the previous block. So the larger `num-anchors` making more overlaps and training loss/acc look unusually better.

### Reproduction

same as https://github.com/sgl-project/SpecForge/issues/465

### Environment

same as https://github.com/sgl-project/SpecForge/issues/465

### LINKED ISSUE #471: [Bug] Fix incorrect padding logic in both backends and wrong target_hidden construction in DFlash training
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/SpecForge/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

**Description:**

I have identified critical issues in the data generation and training logic for DFlash. Specifically, the padding handling in both SGLang and HF backends is incorrect, and the construction of `target_hidden` in the training loop is fundamentally flawed.

### 1. Incorrect Padding Logic in Target Models

**File:** `specforge/modeling/target/dflash_target_model.py`

**Issue:**
There is an inconsistency and error in how padding is handled across different backends:

*   **SGLang Backend (`SGLangDFlashTargetModel`):** It currently pads both `hidden_states` and `input_ids`. Padding `hidden_states` is incorrect and leads to dimension issues downstream.
*   **Hugging Face Backend (`HFDFlashTargetModel`):** It currently performs **no padding** at all. This is also incorrect because `input_ids` must be padded to handle batched training data properly.

**Proposed Fix:**
Both backends should follow the same correct logic:
1.  **Pad `input_ids`**: This is necessary predict next tokens.
2.  **Do NOT pad `hidden_states`**.

### 2. Wrong `block_hidden` Construction in `_build_blocks_from_anchors`

**File:** `specforge/core/dflash.py`

**Issue:**
In `_build_blocks_from_anchors`, the code currently gathers `hidden_states` based on the block indices (`gather_idx`) to create `block_hidden`.

**Problematic Code:**
```python
# specforge/core/dflash.py

def _build_blocks_from_anchors(...):
    # ...
    # This logic gathers hidden states of the tokens INSIDE the block
    block_hidden = torch.gather(
        hidden_states,
        1,
        gather_idx.unsqueeze(-1).expand(-1, -1, hidden_states.size(-1)),
    )
    # ...
```

**Why this is wrong:**
This `block_hidden` is subsequently passed as `target_hidden` to the drafter model. The drafter uses this `target_hidden` as the **conditioning context** (via `self.fc(target_hidden)`).
By gathering the hidden states of the tokens within the block (which are the targets to be predicted), we are destroying the necessary prefix/context information. The `target_hidden` passed to the drafter should **not undergo block-wise gathering**. It should represent the full context features from the target model to correctly condition the generation of the draft block.

### Status
I am actively working on fixing these padding inconsistencies and the hidden state construction logic. A Pull Request will be submitted shortly.

### Reproduction

same as https://github.com/sgl-project/SpecForge/issues/465

### Environment

same as https://github.com/sgl-project/SpecForge/issues/465

### PR DESCRIPTION
**Pull request by xiaomin-D**
## Motivation

Fix two bugs in DFlash training identified in #470 and #471.

### 1. Remove incorrect padding (shift) in SGLang backend (Related: #471)

The `padding()` function in `dflash_target_model.py` is a left-shift operation inherited from Eagle3, where it aligns `hidden_states[t]` with `token[t+1]` for next-token prediction. However, DFlash uses **same-position prediction** (`logits[j]` → `token[j]`), so this shift is incorrect and causes a 1-position misalignment between `input_ids`/`hidden_states` (shifted) and `loss_mask`/`attention_mask` (not shifted).

**Fix**: Remove the `padding()` calls entirely. The HF backend was already correct (no shift).

### 2. Fix cross-block data leak in random-anchor mode (Related: #470)

When `num-anchors` is large, anchor spacing drops below `block_size`, causing blocks to overlap. The old attention mask used **block-id comparison** (`k_ctx < q_b`) to determine context visibility, which allowed a later block to see target hidden states from overlapping positions — a data leak that never occurs during inference.

**Fix**: Change context visibility from block-id comparison to **original position comparison** (`orig_pos[kv] < anchor_pos[query_block]`), ensuring each block can only attend to context strictly before its anchor position, regardless of overlap.

## Related Issues

Fixes #470
Fixes #471

## Checklist

- [x] Format your code according to the [[Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit)](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [[Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci)](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [[Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci)](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [[Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html)](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [[Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html)](https://docs.sglang.ai/references/accuracy_evaluation.html).

### CODE REVIEW

**Comment by xiaomin-D** [comment]
We are currently running experiments to validate the effectiveness of these fixes. Results will be updated here once available.

**Comment by xiaomin-D** [comment]
FYI : https://github.com/sgl-project/SpecForge/pull/473
